v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,194 @@
+		migrationBackoff:    migrationBackoff,
+		dryRun:              dryRun,
+	}
+}
+
+func (c *Cleaner) Run(ctx context.Context) error {
+	list, err := c.apiExtensionsClient.List(ctx, v1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to list CRDs")
+	}
+
+	if list == nil || len(list.Items) == 0 {
+		return errors.New("found 0 results, make sure you have ASO CRDs installed")
+
+	}
+
+	var updated int
+	crdRegexp := regexp.MustCompile(`.*\.azure\.com`)
+	deprecatedVersionRegexp := regexp.MustCompile(`v1alpha1api\d{8}(preview)?(storage)?`)
+	for _, crd := range list.Items {
+		crd := crd
+
+		if !crdRegexp.MatchString(crd.Name) {
+			continue
+		}
+
+		newStoredVersions := removeMatchingStoredVersions(crd.Status.StoredVersions, deprecatedVersionRegexp)
+
+		if len(newStoredVersions) > 0 && len(newStoredVersions) != len(crd.Status.StoredVersions) {
+			klog.Infof("starting cleanup for '%s'", crd.Name)
+			objectsToMigrate, err := c.getObjectsForMigration(ctx, crd, deprecatedVersionRegexp)
+			if err != nil {
+				return err
+			}
+
+			err = c.migrateObjects(ctx, objectsToMigrate)
+			if err != nil {
+				return err
+			}
+
+			err = c.updateStorageVersions(ctx, crd, newStoredVersions)
+			if err != nil {
+				return err
+			}
+
+			updated++
+		} else {
+			klog.Infof("nothing to update for '%s'\n", crd.Name)
+		}
+	}
+
+	if !c.dryRun {
+		klog.Infof("updated %d CRD(s)\n", updated)
+
+	}
+
+	return nil
+}
+
+func (c *Cleaner) updateStorageVersions(
+	ctx context.Context,
+	crd apiextensions.CustomResourceDefinition,
+	newStoredVersions []string) error {
+
+	if c.dryRun {
+		klog.Infof("new storeVersions for '%s' CRD status storedVersions: %s\n", crd.Name, newStoredVersions)
+		return nil
+	}
+
+	crd.Status.StoredVersions = newStoredVersions
+	updatedCrd, err := c.apiExtensionsClient.UpdateStatus(ctx, &crd, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	klog.Infof("updated '%s' CRD status storedVersions to : %s\n", crd.Name, updatedCrd.Status.StoredVersions)
+
+	return nil
+}
+
+func (c *Cleaner) migrateObjects(ctx context.Context, objectsToMigrate *unstructured.UnstructuredList) error {
+	for _, obj := range objectsToMigrate.Items {
+		obj := obj
+		if c.dryRun {
+			// TODO: verbose
+			klog.V(logging.Verbose).Infof("resource '%s' to migrate for kind '%s'", obj.GetName(), obj.GroupVersionKind().Kind)
+			continue
+		}
+
+		err := retry.OnError(c.migrationBackoff, isErrorFatal, func() error { return c.client.Update(ctx, &obj) })
+		if isErrorFatal(err) {
+			return err
+		}
+		klog.V(logging.Verbose).Infof("migrated '%s' for %s\n", obj.GetName(), obj.GroupVersionKind().Kind)
+	}
+
+	return nil
+}
+
+func isErrorFatal(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if apierrors.IsGone(err) { // If resource no longer exists, we don't want to retry
+		return false
+	} else if apierrors.IsConflict(err) { // If resource is already in the state of update, we don't want to retry either
```

Comment:
minor: Possibly expand on why this isn't a problem.

The scenario is:
We try to Update an object, but we get conflict. Conflict means somebody else updated this object between when we listed it and now. BUT because all updates will trigger a conversion to the latest storage version, their update is as good as our update and we know this resource is safe.

Maybe write that up in the comment here somehow just so people understand why this is OK - because at first thought "my update was ignored because somebody else did an update" _seems_ problematic. It's just that we don't actually care about our update (we're not changing anything on the object even)